actions.js

```diff
@@ -1,228 +1,281 @@
+    await contract.functionCall(accountId, 'add_request', {
+      request: {
+        receiver_id: lockupAccountId,
+        actions: [
+          funcCall("termination_withdraw", {receiver_id: accountId})
+        ]
+      }
+    });
+  }
 }
 
-async function vestingTermination(contract, requestKind) {
-    let accountId = contract.accountId;
-    let lockupAccountId = document.querySelector('#lockup-account-id').value;
-    if (!await utils.accountExists(window.near.connection, lockupAccountId)) {
-        alert(`Account ${lockupAccountId} doesn't exist`);
-        return;
-    }
-    const lockupAccount = await window.near.account(lockupAccountId);
-    console.log(`Vesting ${requestKind} for ${lockupAccountId}`);
-    if (requestKind === "terminate_vesting") {
-        await contract.functionCall(accountId, 'add_request', {
-            request: {
-                receiver_id: lockupAccountId,
-                actions: [
-                    funcCall("terminate_vesting", {})
-                ]
-            }
-        });
-    } else if (requestKind === "termination_withdraw") {
-        await contract.functionCall(accountId, 'add_request', {
-            request: {
-                receiver_id: lockupAccountId,
-                actions: [
-                    funcCall("termination_withdraw", { receiver_id: accountId })
-                ]
-            }
-        });
-    }
+function dateToEpoch(date) {
+  return Math.floor(date.getTime() / 1000.0 * 1000000)
+}
+
+async function vestingPrivateTermination(contract, requestKind) {
+  let accountId = contract.accountId;
+  let lockupAccountId = document.querySelector('#lockup-account-id').value;
+  let lockupVestingStartDate = dateToEpoch(new Date(document.querySelector('#lockup-vesting-start-date').value));
+  let lockupVestingEndDate = dateToEpoch(new Date(document.querySelector('#lockup-vesting-end-date').value));
+  let lockupVestingCliffDate = dateToEpoch(new Date(document.querySelector('#lockup-vesting-cliff-date').value));
+  let lockupVestingSalt = document.querySelector('#lockup-vesting-salt').value;
+  if (!await utils.accountExists(window.near.connection, lockupAccountId)) {
+    alert(`Account ${lockupAccountId} doesn't exist`);
+    return;
+  }
+  const lockupAccount = await window.near.account(lockupAccountId);
+  console.log(`Vesting ${requestKind} for ${lockupAccountId}`);
+
+  let publicKeyStr = document.querySelector('#new-key').value;
+  let publicKey = nearAPI.utils.PublicKey.fromString(publicKeyStr);
+
+  const salt = Buffer.from(sha256(Buffer.from(lockupVestingSalt + publicKey)), 'hex').toString('base64');
+
+  let args = {
+    vesting_schedule_with_salt:
+      {
+        vesting_schedule:
+          {
+            start_timestamp: lockupVestingStartDate.toString(),
+            cliff_timestamp: lockupVestingCliffDate.toString(),
+            end_timestamp: lockupVestingEndDate.toString()
+          },
+        salt
+      }
+  }
```

Comment:
it would be good to check here that the hash of this vesting actually matches hash in the contract. Otherwise one can submit a request only to find it's actually wrong after it got confirmed.

Reply:
I suggested it but was decided to skip checks as it is a temporary solution before functionally moved to the lockup UI, and properly addressed, but thanks  for pointing out anyway.